Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java

```diff
@@ -280,6 +280,46 @@ public void onError(RPCError error) {
     }
   }
 
+  /**
+   * Reloaded scripts need to have their breakpoints re-applied; re-set all existing breakpoints.
+   */
+  public void restoreBreakpointsForIsolate(@NotNull final String isolateId, Runnable onFinished) {
+    // Remove all existing VM breakpoints for this isolate.
+    myBreakpointHandler.removeAllVmBreakpoints(isolateId);
+
+    // Re-set existing breakpoints.
+    final Set<XLineBreakpoint<XBreakpointProperties>> xBreakpoints = myBreakpointHandler.getXBreakpoints();
+
+    final AtomicInteger counter = new AtomicInteger(xBreakpoints.size());
+
+    for (final XLineBreakpoint<XBreakpointProperties> xBreakpoint : xBreakpoints) {
+      addBreakpoint(isolateId, xBreakpoint.getSourcePosition(), new VmServiceConsumers.BreakpointConsumerWrapper() {
+        @Override
+        void sourcePositionNotApplicable() {
+          checkDone();
+        }
+
+        @Override
+        public void received(Breakpoint vmBreakpoint) {
+          myBreakpointHandler.vmBreakpointAdded(xBreakpoint, isolateId, vmBreakpoint);
+          checkDone();
+        }
+
+        @Override
+        public void onError(RPCError error) {
+          myBreakpointHandler.breakpointFailed(xBreakpoint);
+          checkDone();
+        }
+
+        private void checkDone() {
+          if (counter.decrementAndGet() == 0 && onFinished != null) {
+            onFinished.run();
```

Comment:
Looks like `onFinished.run()` is never called if `xBreakpoints` is empty. I've extracted common code including the fix for this, no actions needed.